src/main/software/amazon/event/ruler/ComparableNumber.java

```diff
@@ -3,104 +3,122 @@
- * 2-4 times slower for basic mathematical and comparison operations, so we turn to long integer arithmetic.
- * This will need to be modified if we ever need to support larger numbers.
  */
 class ComparableNumber {
-    // Use scientific notation to define the double number directly to avoid losing Precision by calculation
-    // for example 5000 * 1000 *1000 will be wrongly parsed as 7.05032704E8 by computer.
-    static final double HALF_TRILLION = 5E11;
 
-    static final int MAX_LENGTH_IN_BYTES = 16;
-    static final int MAX_DECIMAL_PRECISON = 6;
+    static final int MAX_LENGTH_IN_BYTES = 10;
+    static final int BASE_128_BITMASK = 0x7f; // 127 or 01111111
 
-    public static final BigDecimal TEN_E_SIX = new BigDecimal("1E6"); // to remove decimals
-    public static final long HALF_TRILLION_TEN_E_SIX = new BigDecimal(ComparableNumber.HALF_TRILLION).multiply(TEN_E_SIX).longValueExact();
-
-    private ComparableNumber() {
-    }
+    private ComparableNumber() {}
 
     /**
-     * Generates a hexadecimal string representation of a given decimal string value,
-     * with a maximum precision of 6 decimal places and a range between -5,000,000,000
-     * and 5,000,000,000 (inclusive).
+     * Generates a comparable number string from a given string representation
+     * using numbits representation.
      *
-     * @param str the decimal string value to be converted
-     * @return the hexadecimal string representation of the input value
-     * @throws IllegalArgumentException if the input value has more than 6 decimal places
-     *                                  or is outside the allowed range
+     * @param str the string representation of the number
+     * @return the comparable number string
+     * @throws NumberFormatException if the input isn't a number
+     * @throws IllegalArgumentException if the input isn't a number we can compare
      */
     static String generate(final String str) {
-        final BigDecimal number = JavaBigDecimalParser.parseBigDecimal(str).stripTrailingZeros();
-        if (number.scale() > MAX_DECIMAL_PRECISON) {
-            throw new IllegalArgumentException("Only values upto 6 decimals are supported");
-        }
-
-        final long shiftedBySixDecimals = number.multiply(TEN_E_SIX).longValueExact();
+        final BigDecimal bigDecimal = JavaBigDecimalParser.parseBigDecimal(str);
+        final double doubleValue = bigDecimal.doubleValue();
 
-        // faster than doing bigDecimal comparisons
-        if (shiftedBySixDecimals < -HALF_TRILLION_TEN_E_SIX || shiftedBySixDecimals > HALF_TRILLION_TEN_E_SIX) {
-            throw new IllegalArgumentException("Value must be between " + -ComparableNumber.HALF_TRILLION +
-                    " and " + ComparableNumber.HALF_TRILLION + ", inclusive");
+        // make sure we have the comparable numbers and haven't eaten up decimals values
+        if(Double.isNaN(doubleValue) || Double.isInfinite(doubleValue) ||
```

Comment:
Can this happen? JSON doesn't allow numeric values of NaN and Inf

Reply:
not in JSON but there's interest in supporting other formats so I'm bullet proofing this now than be in for a surprise in future.

Reply:
You can check both in one go based on the bits-representation. With a constant `long NOT_FINITE = 0x7ff0L << 52`: `(bits & NOT_FINITE) == NOT_FINITE` will be true for only NaN and +/- Infinity.

Maybe the JVM automagically already inlines the two calls like that. If not, it could help, here.

Reply:
JVM does inline it. I've left this as it for readability for now.